File: Defender/Guard/OfType.vb
Imports System.Runtime.CompilerServices

Partial Public Module Guard
	''' <summary>
	''' Guard against the argument Not being of type <typeparamref name="T"/>.
	''' </summary>
	''' <typeparam name="T">The type the <paramref name="value"/> must be of.</typeparam>
	''' <param name="value">The value.</param>
	''' <param name="name">The name of the argument.</param>
	<MethodImpl(MethodImplOptions.AggressiveInlining)>
	Public Sub OfType(Of T)(Value As Object, Name As String)
		If TypeOf Value IsNot T Then
			Throw New ArgumentTypeException(Name, GetType(T))
		End If
	End Sub
End Module

File: Defender/Guards/OfType.cs
using System;
using System.Runtime.CompilerServices;
using Defender.Exceptions;

namespace Defender {
	public static partial class Guard {
		/// <summary>
		/// Guard against the argument not being of the type <typeparamref name="T"/>.
		/// </summary>
		/// <typeparam name="T">The type the <paramref name="value"/> must be of.</typeparam>
		/// <param name="value">The value.</param>
		/// <param name="name">The name of the argument.</param>
		[MethodImpl(MethodImplOptions.AggressiveInlining)]
		public static void OfType<T>(Object value, String name) {
			if (value is null || value.GetType() != typeof(T)) {
				throw new ArgumentTypeException(name, typeof(T));
			}
		}
	}
}